### DISCUSSION TITLE
No std::string -> nanobind::str conversion?

### DISCUSSION THREAD

**Discussion by nschloe** [wjakob/nanobind#309] [Q&A] [answered]
When trying to pass a `std::string` to an nb function that accepts an `nb::str`, compilation will fail:

```
note:   no known conversion for argument 1 from ‘std::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘const nanobind::str&’
```

Converting it to a `char*` via `std::string::c_str()` also won't work:

```
note:   no known conversion for argument 1 from ‘const char*’ to ‘const nanobind::str&’
```

String literals work.

Any hints?

**Comment by wjakob** [comment, marked as answer]
`nb::str` has an explicit conversion from `char *` (but not an *implicit* one!). You must cast it like so: `nb::str(std_str.c_str())`. There is no way that there could be a `nb::str` constructor that takes `std::string` — nanobind tries as hard as possible to not include any unnecessary STL headers.

What I would suggest is that you make your own `nb::str` subclass (`nschloe_str` ;)) with all the implicit and explicit constructors you fancy. Since it derives from a builtin wrapper type, nanobind knows how to accept and return it in bindings.

**Reply by nschloe** [reply]
Perhaps that's an item for the [Porting guide](https://nanobind.readthedocs.io/en/latest/porting.html).

**Reply by wjakob** [reply]
Would you be willing to make a PR for such a note?

**Reply by nschloe** [reply]
I can draft something. Perhaps this should be answered together with https://github.com/wjakob/nanobind/discussions/310 under "Strings in Nb".
